### PR TITLE
build: require exact Lua version when using PREFER_LUA

### PR DESCRIPTION
**Pull request by gpanders**
.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -413,7 +413,7 @@ endif()
 option(PREFER_LUA "Prefer Lua over LuaJIT in the nvim executable." OFF)
 
 if(PREFER_LUA)
-  find_package(Lua 5.1 REQUIRED)
+  find_package(Lua 5.1 EXACT REQUIRED)
```
Comment (jamessan) [inline CMakeLists.txt:416]: What's the motivation for this change?  You'll notice that cmake/FindLua.cmake does support building against newer versions of Lua (even though it doesn't make a lot of sense to do so).

Comment (gpanders) [reply CMakeLists.txt:416]: Without this change, if a user has multiple versions of Lua installed (e.g. both 5.1 and 5.4), `FindLua.cmake` will eagerly try to use the "newest" one (5.4 in this case), which fails the build (missing definitions and all that). AFAIK we only support Lua 5.1, so we need to be explicit that we don't want Cmake to do us any favors by trying to give us a newer version if one is found.

Comment (jamessan) [reply CMakeLists.txt:416]: We control `FindLua.cmake`, so if you want to prefer Lua5.1 then switch the order. :) There are distros which insist on packaging neovim with newer versions of Lua (because they don't understand Lua versioning and have dropped Lua 5.1), which is why we have that list in `FindLua.cmake`.

Comment (gpanders) [reply CMakeLists.txt:416]: >There are distros which insist on packaging neovim with newer versions of Lua (because they don't understand Lua versioning and have dropped Lua 5.1)

So I take this to mean that Lua does build (and work) correctly with *some* non-5.1 versions of Lua? It certainly does not work with 5.4 (as I experienced), so instead could we just supply a version range to `find_package`?

Comment (bfredl) [reply CMakeLists.txt:416]: neovim is (deliberately) not compatible with any PuC lua version since (and including) 5.2 on the source level. It might not stop some distros from trying to patch it to compile anyway tho.

Comment (jamessan) [reply CMakeLists.txt:416]: > neovim is (deliberately) not compatible with any PuC lua version since (and including) 5.2 on the source level. It might not stop some distros from trying to patch it to compile anyway tho.

If there are changes needed to _build_ against anything other than Lua 5.1 or LuaJIT, then we should remove the other versions from FindLua.cmake.  If distros already need to patch to make neovim build against different versions, then they can handle that too.

Comment (bfredl) [reply CMakeLists.txt:416]: I thought findLua also affected the use of build time lua, where we _do_ support any lua 5.1+ version (so a distro could use lua5.3+lua5.3-(lpeg+mpack+bit)  as a buildtime dependency and only luajit+luv as the only runtime lua dependencies).

Comment (gpanders) [reply CMakeLists.txt:416]: The change in this PR only affects the version of Lua used at runtime, not build time. The build time Lua is determined elsewhere in the CMakeLists file.

Comment (bfredl) [reply CMakeLists.txt:416]: Sure the PR is fine as is, this was about the suggestion of completely removing 5.4/5.3 support in findLua.cmake, which I don't think we should do.

Comment (jamessan) [reply CMakeLists.txt:416]: This is the only place that uses FindLua.cmake.  FindLua.cmake gives us the include and library paths, so I agree that it's targeting what should be built against for use during runtime.  That would imply that it shouldn't allow unsupported Lua versions.

Now, a bit further down in this file we have the LUA_PRG+lpeg+mpack+bit checks and _those_ only allow using Lua 5.1, 5.2, or LuaJIT.  LUA_PRG is what we use to run the lua scripts and parts of the test suite.

It seems like the version checks we're doing are reversed from what they should be.

Comment (bfredl) [reply CMakeLists.txt:416]: That's weird. I think it was written under the assumption that lua5.3 would be installed/symlinked by distros as `lua` as it was (at the time) the latest version, though it would of course be better to write out `lua5.3` and `lua5.4` (though someone should verify it works) explicityl.

Comment (gpanders) [reply CMakeLists.txt:416]: @jamessan I ran into this again today when building with `-DPREFER_LUA=ON`. Did we come to a consensus on a path forward? Should I modify `FindLua` to remove all versions except 5.1, or do we use the `EXACT` keyword as this PR does?

Comment (jamessan) [reply CMakeLists.txt:416]: As far as I know, https://github.com/neovim/neovim/pull/16633#discussion_r768078718 is the relevant summary.  If that's not disputed, then I'm good with the PR as is.  If you _also_ want to remove the non-5.1 versions from `FindLua`, that's fine with me.
